### PR TITLE
UGENE-7197. Multiline molecule name in pdb-file is parsed incorrectly

### PR DESCRIPTION
**Pull request by evelinabis**
[Issue](https://ugene.dev/tracker/browse/UGENE-7197)

### CODE REVIEW

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp [hunk 1]
```diff
@@ -321,6 +322,24 @@ GUI_TEST_CLASS_DEFINITION(test_7152) {
     CHECK_SET_ERR(bottomRight == "11/40/35", "Bottom right position is wrong: " + bottomRight);
 }
 
+GUI_TEST_CLASS_DEFINITION(test_7197) {
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_7001_8000.cpp]: What is the reason to have GUI (heavy & slow) test for parsers? This should be an XML test.

File: src/corelibs/U2Formats/src/PDBFormat.cpp [hunk 2]
```diff
@@ -280,12 +281,22 @@ void PDBFormat::PDBParser::parseMacromolecularContent(bool firstCompndLine, U2Op
     list
     Details
     */
+
+    // Returns the end-of-name index for `specification`.
+    auto endOfNameInd = [&](const QString &specification) {
+        int index = QRegExp(";\\s*$").indexIn(specification);
```
Comment (yalgaer) [inline src/corelibs/U2Formats/src/PDBFormat.cpp]: Here you create a new reg-ex object on every invocation (may be in loop) which is anti-pattern
Normally you should re-use reg-ex objects. You can make is function static, or a class field.

**Review by yalgaer — CHANGES_REQUESTED** [review]

**Review by smlnkcv — COMMENTED** [review]
Approved

File: src/corelibs/U2Formats/src/PDBFormat.cpp [hunk 3]
```diff
@@ -606,6 +609,15 @@ void PDBFormat::PDBParser::createMolecule(char chainIdentifier, BioStruct3D &bio
 void PDBFormat::PDBParser::updateResidueIndexes(BioStruct3D & /*biostruc*/) {
 }
 
+int PDBFormat::PDBParser::endOfNameInd(const QString &specification) {
+    int index = QRegExp(";\\s*$").indexIn(specification);
```
Comment (yalgaer) [inline src/corelibs/U2Formats/src/PDBFormat.cpp]: Could you please re-use regex instance and avoid new instance creation on every method call?

File: src/corelibs/U2Formats/src/PDBFormat.h [hunk 4]
```diff
@@ -99,6 +103,9 @@ class U2FORMATS_EXPORT PDBFormat : public TextDocumentFormatDeprecated {
         bool seqResContains(char chainIdentier, int residueIndex, char acronym);
         QByteArray getNextSpecLine();
 
+        // Returns the end-of-name index for `specification`.
+        int endOfNameInd(const QString &specification);
```
Comment (yalgaer) [inline src/corelibs/U2Formats/src/PDBFormat.h]: The method is not a simple getter but changes the internal parser state.
Please update the doc and change the name accordingly.

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.h [hunk 5]
```diff
@@ -87,6 +87,17 @@ class GTest_BioStruct3DAtomResidueName : public XmlTest {
     int modelId;
 };
 
+class GTest_BioStruct3DMoleculeName : public XmlTest {
+    Q_OBJECT
+    SIMPLE_XML_TEST_BODY_WITH_FACTORY(GTest_BioStruct3DMoleculeName, "check-biostruct3d-molecule-name");
+
+    ReportResult report();
```
Comment (yalgaer) [inline src/plugins/CoreTests/src/BioStruct3DObjectTests.h]: add override

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.h [hunk 6]
```diff
@@ -87,6 +87,17 @@ class GTest_BioStruct3DAtomResidueName : public XmlTest {
     int modelId;
 };
 
+class GTest_BioStruct3DMoleculeName : public XmlTest {
```
Comment (yalgaer) [inline src/plugins/CoreTests/src/BioStruct3DObjectTests.h:98]: Add class-level doc + usage example

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.h [hunk 7]
```diff
@@ -87,6 +87,17 @@ class GTest_BioStruct3DAtomResidueName : public XmlTest {
...
+    Q_OBJECT
+    SIMPLE_XML_TEST_BODY_WITH_FACTORY(GTest_BioStruct3DMoleculeName, "check-biostruct3d-molecule-name");
+
+    ReportResult report();
+
+    QString objContextName;
+    QString molName;
+    int chainInd;
```
Comment (yalgaer) [inline src/plugins/CoreTests/src/BioStruct3DObjectTests.h]: nit: use fully qualified names: 'Index', not 'Ind' when the length difference is not significant.

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp [hunk 8]
```diff
@@ -391,6 +393,63 @@ Task::ReportResult GTest_BioStruct3DAtomResidueName::report() {
...
+    if (!v.isEmpty()) {
+        bool ok = false;
+        chainInd = v.toInt(&ok);
+        if (!ok) {
+            stateInfo.setError(QString("invalid value type %1, int required").arg(CHAIN_IND_ATTR));
+        }
+    }
+
+    // molecule name
+    v = el.attribute(MOL_NAME_ATTR);
```
Comment (yalgaer) [inline src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp]: nit: use molName=el.attr.. directly (same as for the objContextName above)

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp [hunk 9]
```diff
@@ -391,6 +393,63 @@ Task::ReportResult GTest_BioStruct3DAtomResidueName::report() {
 
     return ReportResult_Finished;
 }
+///////////////////////////////////////////////////////////////////////////////////////////
+
+void GTest_BioStruct3DMoleculeName::init(XMLTestFormat *tf, const QDomElement &el) {
+    Q_UNUSED(tf);
```
Comment (yalgaer) [inline src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp]: Do not declare 'tf' name in the params list and avoid Q_UNUSED

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp [hunk 10]
```diff
@@ -38,6 +38,8 @@ namespace U2 {
 #define X_COORD_ATTR "x"
 #define Y_COORD_ATTR "y"
 #define Z_COORD_ATTR "z"
+#define CHAIN_IND_ATTR "chain-ind"
```
Comment (yalgaer) [inline src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp]: Use full name: "chain-index"

**Review by yalgaer — CHANGES_REQUESTED** [review]

File: src/corelibs/U2Formats/src/PDBFormat.cpp (hunk 3 above)
Comment (DmitriiSukhomlinov) [reply src/corelibs/U2Formats/src/PDBFormat.cpp]: The best way is to create a `static const` variable

File: src/corelibs/U2Formats/src/PDBFormat.h (hunk 4 above)
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Formats/src/PDBFormat.h]: There is no need to short "index" as "ind", rename, please

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp (hunk 9 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp]: I can argue with that. We don't have any agreements about unused variables, actually, so, for example, I prefer wrap them with `Q_UNUSED` too

File: src/corelibs/U2Formats/src/PDBFormat.h (hunk 4 above)
Comment (evelinabis) [reply src/corelibs/U2Formats/src/PDBFormat.h]: The new title looks awful, but the conversation is resolved.

File: src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp (hunk 9 above)
Comment (evelinabis) [reply src/plugins/CoreTests/src/BioStruct3DObjectTests.cpp]: `tf` declaration removed.

**Review by rasputinkirill — APPROVED** [review]

**Review by yalgaer — APPROVED** [review]

**Review by smlnkcv — APPROVED** [review]

**Pull request merged**